# Conditional Probability Simulation

## Example: Soccer Game

You are off to soccer and want to be the goalkeeper but that depends on who will be coaching today:

- with coach Sam, the probability of being the goalkeeper is 0.5
- with coach Alex, the probability of being the goalkeeper is 0.3

You also know that Sam coaches more often – about six out of every ten games (a probability of 0.6).

So, what is the probability that you will be the goalkeeper today?

In [1]:
# import numpy 
import numpy as np

Now, we have to set the number of times we want to simulate this process and initialize all variables with zero.

In [2]:
# how many times we run the simulation
n_runs = 100000

# initialize variables
SAM = 0
SAM_YES = 0
SAM_NO = 0
ALEX = 0
ALEX_YES = 0
ALEX_NO = 0
YES = 0
NO = 0

After we set all the variables to their initial value, we need to write a loop that will simulate the process.

In [3]:
# process simulation
for _ in range(n_runs):
    # Sam (probability of Sam being the coach)
    if np.random.random() < 0.6: 
        SAM += 1

        # YES
        if np.random.random() < 0.5:
            SAM_YES += 1
            YES +=1

        # NO
        else:
            SAM_NO += 1
            NO += 1

    # Alex (probability of Alex being the coach)
    else:
        ALEX += 1

        # YES
        if np.random.random() < 0.3:
            ALEX_YES += 1
            YES += 1

        # NO
        else:
            ALEX_NO += 1
            NO +=1

The simulation above is based on a random number generator (np.random.random), which generates any number from the interval <0,1> with equal probability. If the generated number is less than 0.6 we can say that the coach is Sam (because Sam is the coach in 60% of games). Then we generate another random number and if the value is less than 0.5 (the chance of being the goalkeeper when Sam is coaching is 50%), it means you will be the goalkeeper. In that case, we will increase the value of variables SAM_YES and YES by one. It means that in this simulation run you are the goalkeeper and Sam is the coach. We run this process 100,000 times and check the result.

In [7]:
# create probabilities
P_SAM = SAM / n_runs * 100             # P(Sam)
P_SAM_YES = SAM_YES / n_runs * 100     # P(Sam and Yes)  Note: ie Sam 'intersect' Yes
P_SAM_NO = SAM_NO / n_runs * 100       # P(Sam and No)
P_ALEX = ALEX / n_runs * 100           # P(Alex)
P_ALEX_YES = ALEX_YES / n_runs * 100   # P(Alex and Yes)
P_ALEX_NO = ALEX_NO / n_runs * 100     # P(Alex and No)
P_YES = YES / n_runs * 100             # P(Yes)
P_NO = NO / n_runs * 100               # P(No)

In [8]:
print(f'Sam is the coach: {P_SAM}%')
print(f'Sam is the coach and you are the goalkeeper: {P_SAM_YES}%')
print(f'Sam is the coach and you are not the goalkeeper: {P_SAM_NO}%')
print(f'Alex is the coach: {P_ALEX}%')
print(f'Alex is the coach and you are the goalkeeper: {P_ALEX_YES}%')
print(f'Alex is the coach and you are not the goalkeeper: {P_ALEX_NO}%')
print(f'You are the goalkeeper: {P_YES}%')
print(f'You are not the goalkeeper: {P_NO}')

Sam is the coach: 60.01800000000001%
Sam is the coach and you are the goalkeeper: 30.105999999999998%
Sam is the coach and you are not the goalkeeper: 29.912%
Alex is the coach: 39.982%
Alex is the coach and you are the goalkeeper: 12.029%
Alex is the coach and you are not the goalkeeper: 27.953%
You are the goalkeeper: 42.135%
You are not the goalkeeper: 57.865


As we can see, the probability of being the goalkeeper is around 42%. We can also check the probability of Sam being the coach given that you are the goalkeeper.

In [6]:
# P(Sam | goalkeeper)
P_SAM_GOAL = P_SAM_YES / P_YES
print(P_SAM_GOAL)

0.7145128752818322


We computed this probability as the probability of you being the goalkeeper while Sam is the coach, divided by the total probability that you are the goalkeeper. The probability is around 72%, which means that in 72 out of 100 games in which you are the goalkeeper, your coach is Sam.

In [9]:
P_ALEX_GOAL = P_ALEX_YES / P_YES   # P(Alex INTERSECT Yes) / P(Yes)
print(P_ALEX_GOAL)

0.2854871247181678


### Arithmetical Calculation

P(Sam|Yes) = P(Yes|Sam) * P(Sam) / P(Yes)

P(Yes) = P(Yes|Sam)P(Sam) + P(Yes|Not Sam)P(Not Sam) 
       = 0.5 * 0.6 + 0.3 * 0.4 = 0.42
       
P(Sam|Yes) = 0.5 * 0.6 / 0.42 = 0.71